### PR TITLE
Adding implementation of ServerCallContext and pass to services

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses https://github.com/grpc/grpc-dotnet/issues/35 and https://github.com/grpc/grpc-dotnet/issues/36.

### CODE REVIEW

File: src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+        // TODO(JunTaoLuo, JamesNK): implement this
+        protected override CancellationToken CancellationTokenCore => throw new NotImplementedException();
+
+        // TODO(JunTaoLuo, JamesNK): implement this
+        protected override Metadata ResponseTrailersCore => throw new NotImplementedException();
+
+        protected override Status StatusCore { get; set; }
+
+        protected override WriteOptions WriteOptionsCore { get; set; }
```
Comment (JunTaoLuo) [inline src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs]: I need to hook this up with the HttpContextStreamWriter

**Comment by JunTaoLuo** [comment]
Would be good to see how many more interop tests pass after this change goes in.

File: src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+using System;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.AspNetCore.Http;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal class DefaultServerCallContext : ServerCallContext
```
Comment (JamesNK) [inline src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs]: Lets call it HttpContextServerCallContext

File: src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs (hunk 1 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs]: I've already done it

File: src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        private Lazy<string> _peer;
+
+        internal DefaultServerCallContext(HttpContext httpContext)
+        {
+            HttpContext = httpContext;
+
+            var connection = HttpContext.Connection;
+            if (connection.RemoteIpAddress == null)
+            {
+                _peer = new Lazy<string>();
```
Comment (JamesNK) [inline src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs]: Lazy is overkill. Just make it a string that is default to null and add an if check.

File: src/Grpc.AspNetCore.Server/ServerCallContextExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+using Microsoft.AspNetCore.Http;
+
+namespace Grpc.Core
+{
+    /// <summary>
+    /// Extension methods for ServerCallContext.
+    /// </summary>
+    public static class ServerCallContextExtensions
+    {
+        public static HttpContext GetHttpContext(this ServerCallContext serverCallContext)
```
Comment (JamesNK) [inline src/Grpc.AspNetCore.Server/ServerCallContextExtensions.cs]: XML docs

File: src/Grpc.AspNetCore.Server/Internal/UnaryServerCallHandler.cs [hunk 5]
```diff
@@ -57,13 +57,18 @@ public override async Task HandleCallAsync(HttpContext httpContext)
             // Activate the implementation type via DI.
             var activator = httpContext.RequestServices.GetRequiredService<IGrpcServiceActivator<TService>>();
             var service = activator.Create();
+            var streamWriter = new HttpContextStreamWriter<TResponse>(httpContext, Method.ResponseMarshaller.Serializer);
```
Comment (JamesNK) [inline src/Grpc.AspNetCore.Server/Internal/UnaryServerCallHandler.cs]: Why did you add this?

File: src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs (hunk 1 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs]: What do you mean?

File: src/Grpc.AspNetCore.Server/Internal/UnaryServerCallHandler.cs (hunk 5 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/UnaryServerCallHandler.cs]: Because of WriteOptions. The service can modify the WriteFlags during the call and it can affect how TResponse should be written. For example, the compress flag can be set. Also this standardizes this portion of the logic for all handlers.

File: src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs (hunk 1 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs]: Yeah, I wrote my own temp version of the context. What you've done is fine. I'll rebase the tests I added on top.

File: src/Grpc.AspNetCore.Server/Internal/UnaryServerCallHandler.cs (hunk 5 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/UnaryServerCallHandler.cs]: Ok. Then we should have a write helper method that takes the serializer, write options and message. The StreamWriter and this place can use it. That avoids us allocating the stream writer here.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                _peer = new Lazy<string>($"{(connection.RemoteIpAddress.AddressFamily == AddressFamily.InterNetwork ? "ipv4" : "ipv6")}:{connection.RemoteIpAddress}:{connection.RemotePort}");
+            }
+
+            _requestHeaders = new Lazy<Metadata>(() =>
+            {
+                var metadata = new Metadata();
+
+                foreach (var header in httpContext.Request.Headers)
+                {
+                    if (!header.Key.StartsWith(':'))
```
Comment (JunTaoLuo) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: @jtattermusch looks like there's some discrepancies between how headers are handled and I just want to call them out:
1. Metadata converts header names to lower-case whereas ASP.NET Core does not
2. ASP.NET Core exposes pseudo-headers (:path for example) in the header collection whereas Metadata doesn't allow this.
3. We do not store binary header values. We parse them as UTF-8.

Are we okay with these behavioural differences?

File: src/Grpc.AspNetCore.Server/Internal/ClientStreamingServerCallHandler.cs [hunk 7]
```diff
@@ -51,16 +51,17 @@ public override async Task HandleCallAsync(HttpContext httpContext)
             // Activate the implementation type via DI.
             var activator = httpContext.RequestServices.GetRequiredService<IGrpcServiceActivator<TService>>();
             var service = activator.Create();
+            var streamWriter = new HttpContextStreamWriter<TResponse>(httpContext, Method.ResponseMarshaller.Serializer);
```
Comment (JamesNK) [inline src/Grpc.AspNetCore.Server/Internal/ClientStreamingServerCallHandler.cs]: Remove here as well

File: src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs (hunk 1 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs]: What does your implementation look like? I found my way of doing it not quite ideal.

File: src/Grpc.AspNetCore.Server/Internal/IWriteOptions.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+#endregion
+
+using Grpc.Core;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal interface IWriteOptions
+    {
+        WriteOptions WriteOptions { get; set; }
```
Comment (JamesNK) [inline src/Grpc.AspNetCore.Server/Internal/IWriteOptions.cs]: Get rid of this. Just pass the ServerCallContext to the stream writer.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal class HttpContextServerCallContext : ServerCallContext
+    {
+        private readonly IWriteOptions _writeOptions;
+        private readonly Lazy<Metadata> _requestHeaders;
+        private Lazy<string> _peer;
+
+        internal HttpContextServerCallContext(HttpContext httpContext, IWriteOptions writeOptions)
```
Comment (JamesNK) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Pass the context to the stream writer instead of the stream writer to the context.

File: src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs (hunk 3 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs]: I was hoping to avoid synthesizing this string unless it's called.

File: src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs (hunk 1 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/DefaultServerCallContext.cs]: https://github.com/JamesNK/grpc-dotnet/commit/ee097a4637812ceb695f38545bac21adee9948a0

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        {
+            HttpContext = httpContext;
+
+            var connection = HttpContext.Connection;
+            if (connection.RemoteIpAddress != null)
+            {
+                PeerCore = $"{(connection.RemoteIpAddress.AddressFamily == AddressFamily.InterNetwork ? "ipv4" : "ipv6")}:{connection.RemoteIpAddress}:{connection.RemotePort}";
+            }
+
+            _requestHeaders = new Lazy<Metadata>(() =>
```
Comment (JamesNK) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Don't use a lazy. Just move this logic to the getter.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        // TODO(JunTaoLuo, JamesNK): implement this
+        protected override DateTime DeadlineCore => throw new NotImplementedException();
+
+        protected override Metadata RequestHeadersCore
+        {
+            get
+            {
+                if (_requestHeaders == null)
+                {
+                    _requestHeaders = new Metadata();
```
Comment (JunTaoLuo) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: In general I'd be concerned about duplicating collections. I need to verify that the header dictionary is immutable at this point.

Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Similar issue with ResponseTrailers.

Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: From what I can tell the HttpContext related types aren't designed to be thread safe, e.g. https://github.com/aspnet/AspNetCore/blob/7bd5297cfa672c797e2fcbfffa6dc048699abd32/src/Http/Http/src/Features/ResponseCookiesFeature.cs#L55-L67

File: src/Grpc.AspNetCore.Server/Internal/HttpResponseExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+                    // TODO(juntaoluo): What about binary trailers
+                    if (!trailer.IsBinary)
+                    {
+                        httpResponse.AppendTrailer(trailer.Key, trailer.Value);
+                    }
+                }
+            }
+
+            // Append status trailers, these overwrite any existing status trailers set via ServerCallContext.ResponseTrailers
+            httpResponse.AppendTrailer(GrpcProtocolConstants.StatusTrailer, context.Status.StatusCode.ToString("D"));
```
Comment (JamesNK) [inline src/Grpc.AspNetCore.Server/Internal/HttpResponseExtensions.cs]: I believe ToString("D") will allocation. We could have a method with a switch statement that returns the appropriate string directly. default can fall back to ToString("D").

File: src/Grpc.AspNetCore.Server/Internal/HttpResponseExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+                {
+                    // TODO(juntaoluo): What about binary trailers
+                    if (!trailer.IsBinary)
+                    {
+                        httpResponse.AppendTrailer(trailer.Key, trailer.Value);
+                    }
+                }
+            }
+
+            // Append status trailers, these overwrite any existing status trailers set via ServerCallContext.ResponseTrailers
```
Comment (JunTaoLuo) [inline src/Grpc.AspNetCore.Server/Internal/HttpResponseExtensions.cs]: @jtattermusch I tested the behaviour with the c-core server and this is what I observed. Does that sound correct to you?

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 11 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: It's not really thread safety, I was more concerned that operations on one collection will not be propagated to the other. I think I can update the implementation to make sure they are in sync though.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+                {
+                    if (!entry.IsBinary)
+                    {
+                        // TODO(juntaoluo): what about binary headers?
+                        HttpContext.Response.Headers[entry.Key] = entry.Value;
+                    }
+                }
+            }
+
+            return HttpContext.Response.Body.FlushAsync();
```
Comment (jtattermusch) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Is this the right way to force sending the response headers? (we need to support a scenario where server responds with headers, but response messages are sent much later).

Btw, what happens if I've already sent some payload (= a message) in the body?  In that case this method should give an error.

Also, in Grpc.Core, you can only write the response headers once. Second attempt gives an error. Does this behave the same?

File: src/Grpc.AspNetCore.Server/Internal/HttpResponseExtensions.cs (hunk 13 above)
Comment (jtattermusch) [reply src/Grpc.AspNetCore.Server/Internal/HttpResponseExtensions.cs]: I think you're right. setting the status and details "manually" via the Metadata class is actually something that no one really does.  Btw, in c-core there's actually a different API through which we pass the status of the call and I'm pretty sure that overwrites the metadata entries with the same header name.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 11 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Modifying the request headers on the server seems kind of odd. The amount of effort to keep them in sync doesn't seem like it would be worth that scenario.

File: src/Grpc.AspNetCore.Server/Internal/ServerStreamingServerCallHandler.cs [hunk 15]
```diff
@@ -55,12 +55,19 @@ public override async Task HandleCallAsync(HttpContext httpContext)
             // Activate the implementation type via DI.
             var activator = httpContext.RequestServices.GetRequiredService<IGrpcServiceActivator<TService>>();
             var service = activator.Create();
+            var serverCallContext = new HttpContextServerCallContext(httpContext);
+            var streamWriter = new HttpContextStreamWriter<TResponse>(serverCallContext, Method.ResponseMarshaller.Serializer);
 
-            await _invoker(service, request, new HttpContextStreamWriter<TResponse>(httpContext, Method.ResponseMarshaller.Serializer), null);
+            await _invoker(
```
Comment (jtattermusch) [inline src/Grpc.AspNetCore.Server/Internal/ServerStreamingServerCallHandler.cs]: one thing that's missing here is to handle RpcException thrown by user's code.
That's an alternative way to tell the handling stack to set the status and trailers that are set in the exception.

See https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/Internal/ServerCallHandler.cs#L296.

Same of other call types.

I think it's fine to address in a followup PR.

File: src/Grpc.AspNetCore.Server/ServerCallContextExtensions.cs [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    /// Extension methods for ServerCallContext.
+    /// </summary>
+    public static class ServerCallContextExtensions
+    {
+        /// <summary>
+        /// Retrieve the <see cref="HttpContext"/> from a <see cref="ServerCallContext"/> if possible.
+        /// </summary>
+        /// <param name="serverCallContext">The <see cref="ServerCallContext"/> to extract from.</param>
+        /// <returns>The extracted <see cref="HttpContext"/>. If it cannot be extracted, null will be returned.</returns>
+        public static HttpContext GetHttpContext(this ServerCallContext serverCallContext)
```
Comment (jtattermusch) [inline src/Grpc.AspNetCore.Server/ServerCallContextExtensions.cs]: nit: add a note that this is for power users (and recommend read-only access)?   trying to change stuff in the httpcontext while the grpc layer potentially accesses the same fields can easily break things.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 6 above)
Comment (jtattermusch) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: > @jtattermusch looks like there's some discrepancies between how headers are handled and I just want to call them out:
> 
> 1. Metadata converts header names to lower-case whereas ASP.NET Core does not

not sure where that would be a problem. a grpc app would  access the metadata through ServerCallContext where all the entries seems to have the right casing. On the wire everything will be lowercase because it's required by HTTP/2 spec.

> 2. ASP.NET Core exposes pseudo-headers (:path for example) in the header collection whereas Metadata doesn't allow this.

Looks like that's already addressed by hiding the pseudo headers?

> 3. We do not store binary header values. We parse them as UTF-8.

When writing, the "-bin" metadata entries should just be base64 encoded.

when reading the request headers, we can recognize the "-bin" suffix and base64 decode the value when creating Metadata.Entry.

> 
> Are we okay with these behavioural differences?

File: test/Grpc.AspNetCore.Server.Tests/DefaultServerCallContextTests.cs [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+using Microsoft.Extensions.Primitives;
+using NUnit.Framework;
+
+namespace Grpc.AspNetCore.Server.Tests
+{
+    [TestFixture]
+    public class DefaultServerCallContextTests
```
Comment (jtattermusch) [inline test/Grpc.AspNetCore.Server.Tests/DefaultServerCallContextTests.cs]: nit: rename class to `HttpContextServerCallContextTests`?

File: src/Grpc.AspNetCore.Server/Internal/ServerStreamingServerCallHandler.cs (hunk 15 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/ServerStreamingServerCallHandler.cs]: Yes I was thinking that work belongs in https://github.com/grpc/grpc-dotnet/issues/47.

File: src/Grpc.AspNetCore.Server/ServerCallContextExtensions.cs (hunk 16 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/ServerCallContextExtensions.cs]: I'd add a remarks section and say that this method provides access to the underlying HttpContext and that HttpContext is only available when using gRPC hosted on ASP.NET Core.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 11 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: I think this is safe albeit not quite efficient.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 6 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Right, I'm trying to replicate the gRPC behaviours here as much as possible but I just want to highlight them so we are aware of these differences. The only remaining point is the binary headers. We don't support these. As in, if we receive binary headers that cannot be parsed as ASCII or UTF-8. Kestrel will throw and we reject the request and sends a 400 response. I'll file a follow up issue to discuss this further as this requires a significant amount of work to support on our side.

Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Filed https://github.com/grpc/grpc-dotnet/issues/71

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+        protected override Task WriteResponseHeadersAsyncCore(Metadata responseHeaders)
+        {
+            if (responseHeaders != null)
+            {
+                foreach (var entry in responseHeaders)
+                {
+                    if (!entry.IsBinary)
+                    {
+                        // TODO(juntaoluo): what about binary headers?
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Let's discuss this in the issue.

[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: I see, I'll address this in a separate PR.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 14 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: There's a few discrepancies in behaviour but this is the closest we can get with the current APIs available.

Yes, this allows the user to send the headers without sending any body since the Body may be empty. The current implementation is that any write to the body is flushed which automatically sends headers if they have not been sent yet. There's no scenario where the body can be sent without the headers. Currently flush is going to no-op if the headers has already been sent. How important is it to throw an error instead? I need to think about how this might be achieved on our side as it's not immediately obvious how this can be achieved.

**Comment by JunTaoLuo** [comment]
Thanks for the comments! I'm going to merge this as is to unblock other changes but I'll address a few more feedback items in followup PRs.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 14 above)
Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs]: Actually I lied. there's a super easy way to do this. `HttpContext.Response.HasStarted` I'll still addresss this in a followup PR though.
